File: okone/src/main/java/com/cdh/okone/connection/PreConnectRunnable.java
package com.cdh.okone.connection;

import com.cdh.okone.connection.callback.PreConnectCallback;

import java.io.IOException;
import java.lang.reflect.Field;
import java.lang.reflect.Method;
import java.util.List;
import java.util.concurrent.ConcurrentLinkedQueue;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.SSLSocketFactory;

import okhttp3.Address;
import okhttp3.CertificatePinner;
import okhttp3.EventListener;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.Route;
import okhttp3.internal.connection.RealConnection;
import okhttp3.internal.connection.RealConnectionPool;
import okhttp3.internal.connection.RouteSelector;

/**
 * Created by chidehang on 2020/12/22
 */
public class PreConnectRunnable implements Runnable {

    private static final String THREAD_NAME_PREFIX = "pre-connect-";

    private OkHttpClient mClient;
    private String mUrl;
    private PreConnectCallback mPreConnectCallback;

    public PreConnectRunnable(OkHttpClient client, String url, PreConnectCallback preConnectCallback) {
        this.mClient = client;
        this.mUrl = url;
        this.mPreConnectCallback = preConnectCallback;
    }

    @Override
    public void run() {
        String oldName = Thread.currentThread().getName();
        Thread.currentThread().setName(THREAD_NAME_PREFIX + mUrl);
        innerRun();
        Thread.currentThread().setName(oldName);
    }

    private void innerRun() {
        try {
            HttpUrl httpUrl = createHttpUrl(mUrl);
            if (httpUrl == null) {
                callConnectFailed(mPreConnectCallback, new IllegalArgumentException("unexpected url: " + mUrl));
                return;
            }

            Address address = createAddress(mClient, httpUrl);
            List<Route> routes = selectRoutes(mClient, address);
            if (routes == null || routes.size() == 0) {
                callConnectFailed(mPreConnectCallback, new IOException("No route available."));
                return;
            }

            RealConnectionPool realConnectionPool = findRealConnectionPool(mClient);
            if (hasPooledConnection(realConnectionPool, address, routes, false)) {
                callConnectFailed(mPreConnectCallback, new IllegalStateException("There is already a connection with the same address."));
                return;
            }

            Route route = routes.get(0);
            RealConnection connection = new RealConnection(realConnectionPool, route);
            // 开始连接，如果失败，内部将抛异常
            connection.connect(
                    mClient.connectTimeoutMillis(),
                    mClient.readTimeoutMillis(),
                    mClient.writeTimeoutMillis(),
                    mClient.pingIntervalMillis(),
                    false,
                    null,
                    EventListener.NONE
            );
            mClient.getRouteDatabase().connected(connection.route());

            if (hasPooledConnection(realConnectionPool, address, routes, true)) {
                try {
                    connection.socket().close();
                } catch (Throwable t) {
                }
                callConnectFailed(mPreConnectCallback, new IllegalStateException("There is already a connection with the same address."));
            }

            synchronized (connection) {
                realConnectionPool.put(connection);
            }

            callConnectCompleted(mPreConnectCallback, mUrl);
        } catch (Throwable t) {
            t.printStackTrace();
            callConnectFailed(mPreConnectCallback, t);
        }
    }

    private static HttpUrl createHttpUrl(String url) {
        if (url.regionMatches(true, 0, "ws:", 0, 3)) {
            url = "http:" + url.substring(3);
        } else if (url.regionMatches(true, 0, "wss:", 0, 4)) {
            url = "https:" + url.substring(4);
        }
        return HttpUrl.parse(url);
    }

    private static Address createAddress(OkHttpClient client, HttpUrl url) {
        SSLSocketFactory sslSocketFactory = null;
        HostnameVerifier hostnameVerifier = null;
        CertificatePinner certificatePinner = null;
        if (url.isHttps()) {
            sslSocketFactory = client.sslSocketFactory();
            hostnameVerifier = client.hostnameVerifier();
            certificatePinner = client.certificatePinner();
        }
        return new Address(
                url.host(),
                url.port(),
                client.dns(),
                client.socketFactory(),
                sslSocketFactory,
                hostnameVerifier,
                certificatePinner,
                client.proxyAuthenticator(),
                client.proxy(),
                client.protocols(),
                client.connectionSpecs(),
                client.proxySelector()
        );
    }

    private static List<Route> selectRoutes(OkHttpClient client, Address address) throws IOException {
        RouteSelector routeSelector = new RouteSelector(address, client.getRouteDatabase(), null, EventListener.NONE);
        RouteSelector.Selection selection = routeSelector.hasNext()? routeSelector.next() : null;
        return selection == null? null : selection.getRoutes();
    }

    /**
     * 检查是否已经有缓存的相同地址的连接
     */
    private static boolean hasPooledConnection(RealConnectionPool realConnectionPool, Address address, List<Route> routes, boolean requireMultiplexed) {
        try {
            Field connectionsField = realConnectionPool.getClass().getField("connections");
            connectionsField.setAccessible(true);
            ConcurrentLinkedQueue<RealConnection> connections = (ConcurrentLinkedQueue<RealConnection>) connectionsField.get(realConnectionPool);

            // 遍历现存connection
            for (RealConnection connection : connections) {
                synchronized (connection) {
                    if (requireMultiplexed) {
                        Field field = connection.getClass().getDeclaredField("isMultiplexed");
                        field.setAccessible(true);
                        boolean value = (boolean) field.get(connection);
                        if (!value) {
                            continue;
                        }
                    }

                    Method method = connection.getClass().getDeclaredMethod("isEligible", Address.class, List.class);
                    method.setAccessible(true);
                    boolean result = (boolean) method.invoke(connection, address, routes);
                    if (!result) {
                        continue;
                    }

                    // 已经存在connection
                    return true;
                }
            }
            return false;
        } catch (Exception e) {
            e.printStackTrace();
            throw new UnsupportedOperationException("Does not support the current version of okhttp.");
        }
    }

    private static RealConnectionPool findRealConnectionPool(OkHttpClient client) throws NoSuchFieldException, IllegalAccessException {
        Field delegateField = client.connectionPool().getClass().getDeclaredField("delegate");
        delegateField.setAccessible(true);
        RealConnectionPool delegate = (RealConnectionPool) delegateField.get(client.connectionPool());
        return delegate;
    }

    private static void callConnectCompleted(PreConnectCallback callback, String url) {
        if (callback != null) {
            callback.connectCompleted(url);
        }
    }

    private static void callConnectFailed(PreConnectCallback callback, Throwable t) {
        if (callback != null) {
            callback.connectFailed(t);
        }
    }
}

File: okone/src/main/java/com/cdh/okone/connection/PreConnectRunnable.kt
package com.cdh.okone.connection

import com.cdh.okone.connection.callback.PreConnectCallback
import com.cdh.okone.util.LogUtils
import com.cdh.okone.util.LogUtils.d
import okhttp3.*
import okhttp3.internal.connection.RealConnection
import okhttp3.internal.connection.RealConnectionPool
import okhttp3.internal.connection.RouteSelector
import java.io.IOException
import java.util.concurrent.ConcurrentLinkedQueue
import javax.net.ssl.HostnameVerifier
import javax.net.ssl.SSLSocketFactory

/**
 * Created by chidehang on 2021/2/10
 */
class PreConnectRunnable(
        val mClient: OkHttpClient,
        val mUrl: String,
        val mPreConnectCallback: PreConnectCallback?
) : Runnable {

    override fun run() {
        val oldName = Thread.currentThread().name
        Thread.currentThread().name = THREAD_NAME_PREFIX + mUrl
        innerRun()
        Thread.currentThread().name = oldName
    }

    private fun innerRun() {
        if (LogUtils.isEnabled) d(TAG, "PreConnectRunnable#innerRun() called")
        try {
            val httpUrl = createHttpUrl(mUrl)
            if (httpUrl == null) {
                callConnectFailed(mPreConnectCallback, IllegalArgumentException("unexpected url: $mUrl"))
                return
            }

            val address = createAddress(mClient, httpUrl)
            val routes = selectRoutes(mClient, address)
            if (routes == null || routes.isEmpty()) {
                callConnectFailed(mPreConnectCallback, IOException("No route available."))
                return
            }

            val realConnectionPool = findRealConnectionPool(mClient)
            if (hasPooledConnection(realConnectionPool, address, routes, false)) {
                callConnectFailed(mPreConnectCallback, IllegalStateException("There is already a connection with the same address.[1]"))
                return
            }

            val route = routes[0]
            val connection = RealConnection(realConnectionPool, route)
            // 开始连接，如果失败，内部将抛异常
            connection.connect(
                    mClient.connectTimeoutMillis(),
                    mClient.readTimeoutMillis(),
                    mClient.writeTimeoutMillis(),
                    mClient.pingIntervalMillis(),
                    false,
                    BuildConnectionProcessor.NONE_CALL,
                    EventListener.NONE
            )
            mClient.routeDatabase.connected(connection.route())

            if (hasPooledConnection(realConnectionPool, address, routes, true)) {
                try {
                    connection.socket().close()
                } catch (t: Throwable) {
                }
                callConnectFailed(mPreConnectCallback, java.lang.IllegalStateException("There is already a connection with the same address.[2]"))
                return
            }

            synchronized(connection) { realConnectionPool.put(connection) }

            callConnectCompleted(mPreConnectCallback, mUrl)
        } catch (t: Throwable) {
            t.printStackTrace();
            callConnectFailed(mPreConnectCallback, t);
        }
    }

    companion object {
        private const val TAG = "PreConnectRunnable"
        private const val THREAD_NAME_PREFIX = "pre-connect-"

        private fun createHttpUrl(url: String): HttpUrl? {
            var url = url
            if (url.regionMatches(0, "ws:", 0, 3, ignoreCase = true)) {
                url = "http:" + url.substring(3)
            } else if (url.regionMatches(0, "wss:", 0, 4, ignoreCase = true)) {
                url = "https:" + url.substring(4)
            }
            return HttpUrl.parse(url)
        }

        private fun createAddress(client: OkHttpClient, url: HttpUrl): Address {
            var sslSocketFactory: SSLSocketFactory? = null
            var hostnameVerifier: HostnameVerifier? = null
            var certificatePinner: CertificatePinner? = null
            if (url.isHttps) {
                sslSocketFactory = client.sslSocketFactory()
                hostnameVerifier = client.hostnameVerifier()
                certificatePinner = client.certificatePinner()
            }
            return Address(
                    url.host(),
                    url.port(),
                    client.dns(),
                    client.socketFactory(),
                    sslSocketFactory,
                    hostnameVerifier,
                    certificatePinner,
                    client.proxyAuthenticator(),
                    client.proxy(),
                    client.protocols(),
                    client.connectionSpecs(),
                    client.proxySelector()
            )
        }

        @Throws(IOException::class)
        private fun selectRoutes(client: OkHttpClient, address: Address): List<Route>? {
            val routeSelector = RouteSelector(address, client.routeDatabase, BuildConnectionProcessor.NONE_CALL, EventListener.NONE)
            val selection = if (routeSelector.hasNext()) routeSelector.next() else null
            return selection?.routes
        }

        /**
         * 检查是否已经有缓存的相同地址的连接
         */
        private fun hasPooledConnection(realConnectionPool: RealConnectionPool, address: Address, routes: List<Route>, requireMultiplexed: Boolean): Boolean {
            return try {
                val connectionsField = realConnectionPool.javaClass.getDeclaredField("connections")
                connectionsField.isAccessible = true
                val connections = connectionsField[realConnectionPool] as ConcurrentLinkedQueue<RealConnection>

                // 遍历现存connection
                for (connection in connections) {
                    synchronized(connection) {
                        var dup = true

                        if (requireMultiplexed) {
                            val isMultiplexed = connection.javaClass.getDeclaredMethod("isMultiplexed\$okhttp")
                            isMultiplexed.isAccessible = true
                            val value = isMultiplexed.invoke(connection) as Boolean
                            if (!value) {
                                dup = false
                            }
                        }

                        if (dup) {
                            val isEligible = connection.javaClass.getDeclaredMethod("isEligible\$okhttp", Address::class.java, MutableList::class.java)
                            isEligible.isAccessible = true
                            val result = isEligible.invoke(connection, address, routes) as Boolean
                            if (!result) {
                                dup = false
                            }
                        }

                        if (dup) {
                            // 已经存在connection
                            return true
                        }
                    }
                }
                false
            } catch (e: Exception) {
                e.printStackTrace()
                throw UnsupportedOperationException("Does not support the current version of okhttp.")
            }
        }

        @Throws(NoSuchFieldException::class, IllegalAccessException::class)
        private fun findRealConnectionPool(client: OkHttpClient): RealConnectionPool {
            val delegateField = client.connectionPool().javaClass.getDeclaredField("delegate")
            delegateField.isAccessible = true
            return delegateField[client.connectionPool()] as RealConnectionPool
        }

        private fun callConnectCompleted(callback: PreConnectCallback?, url: String) {
            if (LogUtils.isEnabled) d(TAG, "callConnectCompleted() called with: callback = [$callback], url = [$url]")
            callback?.connectCompleted(url)
        }

        private fun callConnectFailed(callback: PreConnectCallback?, t: Throwable) {
            if (LogUtils.isEnabled) d(TAG, "callConnectFailed() called with: callback = [$callback], t = [$t]")
            callback?.connectFailed(t)
        }
    }
}